File: pubnub-java/src/main/java/com/pubnub/api/endpoints/presence/SetState.java
package com.pubnub.api.endpoints.presence;

import com.pubnub.api.PubNubException;
import com.pubnub.api.endpoints.Endpoint;
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction;
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction;
import com.pubnub.api.models.consumer.presence.PNSetStateResult;
import com.pubnub.internal.PubNubImpl;
import com.pubnub.internal.endpoints.presence.Heartbeat;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.List;


@Accessors(chain = true, fluent = true)
public class SetState extends Endpoint<PNSetStateResult> {
    @Setter
    private List<String> channels = new ArrayList<>();
    @Setter
    private List<String> channelGroups = new ArrayList<>();
    @Setter
    private Object state;
    @Setter
    private String uuid = pubnub.getConfiguration().getUuid();
    @Setter
    private boolean withHeartbeat;


    public SetState(PubNubImpl pubnub) {
        super(pubnub);
    }

    @Override
    protected void validateParams() throws PubNubException {
        if (withHeartbeat) {
            if (uuid != null && !uuid.equals(pubnub.getConfiguration().getUserId().getValue())) {
                // TODO potentially bring back PubNub exception
                throw new IllegalStateException("UserId can't be different from UserId in configuration when flag withHeartbeat is set to true");
            }
        }
    }

    @Override
    protected ExtendedRemoteAction<PNSetStateResult> createAction() {
        if (!withHeartbeat) {
            return pubnub.setPresenceState(
                    channels,
                    channelGroups,
                    state,
                    uuid
            );
        } else {
            return new MappingRemoteAction<>(
                    new Heartbeat(pubnub, channels, channelGroups, state), aBoolean -> new PNSetStateResult(pubnub.getMapper().toJsonTree(state)));
        }
    }
}

File: pubnub-kotlin-core/src/main/kotlin/com/pubnub/internal/endpoints/presence/SetState.kt
package com.pubnub.internal.endpoints.presence

import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.models.consumer.presence.PNSetStateResult
import com.pubnub.api.retry.RetryableEndpointGroup
import com.pubnub.internal.Endpoint
import com.pubnub.internal.PubNubImpl
import com.pubnub.internal.models.server.Envelope
import com.pubnub.internal.presence.eventengine.data.PresenceData
import com.pubnub.internal.toCsv
import retrofit2.Call
import retrofit2.Response

/**
 * @see [PubNubImpl.setPresenceState]
 */
class SetState internal constructor(
    pubnub: PubNubImpl,
    override val channels: List<String>,
    override val channelGroups: List<String>,
    override val state: Any,
    override val uuid: String = pubnub.configuration.userId.value,
    private val presenceData: PresenceData
) : Endpoint<Envelope<JsonElement>, PNSetStateResult>(pubnub), ISetState {

    override fun getAffectedChannels() = channels
    override fun getAffectedChannelGroups() = channelGroups

    override fun validateParams() {
        super.validateParams()
        if (channels.isEmpty() && channelGroups.isEmpty())
            throw PubNubException(PubNubError.CHANNEL_AND_GROUP_MISSING)
    }

    override fun doWork(queryParams: HashMap<String, String>): Call<Envelope<JsonElement>> {
        if (uuid == pubnub.configuration.userId.value) {
            val stateCopy = pubnub.mapper.fromJson(pubnub.mapper.toJson(state), JsonElement::class.java)
            presenceData.channelStates.putAll(channels.associateWith { stateCopy })
        }

        addQueryParams(queryParams)

        return pubnub.retrofitManager.presenceService.setState(
            pubnub.configuration.subscribeKey,
            channels.toCsv(),
            uuid,
            queryParams
        )
    }

    override fun createResponse(input: Response<Envelope<JsonElement>>): PNSetStateResult {
        if (input.body()!!.payload!! is JsonNull) {
            throw PubNubException(PubNubError.PARSING_ERROR)
        }
        return PNSetStateResult(state = input.body()!!.payload!!)
    }

    override fun operationType() = PNOperationType.PNSetStateOperation

    override fun getEndpointGroupName(): RetryableEndpointGroup = RetryableEndpointGroup.PRESENCE

    private fun addQueryParams(queryParams: MutableMap<String, String>) {
        if (channelGroups.isNotEmpty()) queryParams["channel-group"] = channelGroups.toCsv()
        queryParams["state"] = pubnub.mapper.toJson(state)
    }
}
